lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -649,93 +659,108 @@ def _post(self, params: ApiParams, _) -> Response:
             API_METHOD.POST, ParamType.DATASET, params
         ).value
 
-        # Build an Elasticsearch instance to manage the bulk update
-        elastic = Elasticsearch(self.elastic_uri)
-        self.logger.info("Elasticsearch {} [{}]", elastic, VERSION)
+        if self.require_stable and dataset.state.mutating:
```

Comment:
Approaching this another way...when we call `dataset.advance(States.DELETING)` in `DatasetsDelete.generate_actions()` we _should_ get a `DatasetBadStateTransition` if the dataset is currently in a mutating state...or, that is, we _would_ if the transitions table didn't permit that transition....  That seems like a better approach to addressing this problem than adding lots of state, here.  Did you consider doing that?

Of course, if you go that way, you'll presumably need to create a new state for publishing...but, given the fact that publishing prompts a bulk ES transaction, perhaps a separate state is warranted _anyway_....

Reply:
Both publish and delete require a stable state: `UPLOADED` or `INDEXED`. (Or, technically, `DELETED`, but the dataset doesn't exist at that point.)

I don't want "publish" to be a state; it's not, it's just an operation. So the state table doesn't help us here. Whether I should hack further at the state table is mostly unrelated, and I'm not tackling that in this context.

Reply:
> I don't want "publish" to be a state; it's not, it's just an operation. So the state table doesn't help us here.

We wouldn't need a "publish" state, just a "publishing" state (with transitions in from "indexed" and back out to "indexed"), and I think that would allow the state table to cover the problem without most of the changes proposed in this PR.  (Alternatively, we could use the "indexing" state while publishing, which wouldn't be that big of a stretch.)

I wouldn't propose hacking the state table in this PR, unless it's part of the solution, which I think might be a more elegant approach.

Reply:
In theory it's an interesting notion, and in theory that makes it easy to do. In practice, this would be vastly more complicated than I want to even consider here.

The `dataset.state` is a *tracking* mechanism, not a *control* or synchronization mechanism. The control/synchronization side used to be the file system symlinks. We're moving towards a message passing model, and my `Sync` object got rid of the symlinks and moved us a step closer, but we're not "there".

If I add `States.PUBLISHING`, there's nothing to keep the indexer from pulling a dataset in the "publishing" state, and simply failing the `Dataset.advance` to `INDEXING`. What we'd need instead is additional logic to recognize that there's no *problem*, but that the dataset simply isn't available for indexing.

What that really would mean is to drain the `server.operations` list, or possibly just have `Sync.next` omit any mutating dataset until the next round. Maybe that wouldn't be too bad, but I'm still not convinced I want to boil that lake here.

Reply:
> The `dataset.state` is a _tracking_ mechanism, not a _control_ or synchronization mechanism.

Understood.  But, at least in the immediate term, I think we're going to lean on it as though it were a control mechanism.  That is, while I'm pretty sure that there is nothing that we can do to prevent two simultaneous requests, one a "delete" and one a "publish", from proceeding if the dataset is already in `INDEXED` state (unless the Server is single threaded...which, perhaps it is...), if one arrives slightly before the other, so that the second finds the dataset in `DELETING` or `PUBLISHING` state, then we can report the conflict and avoid the collision problem.

> If I add `States.PUBLISHING`, there's nothing to keep the indexer from pulling a dataset in the "publishing" state

In order for a dataset to be in the `PUBLISHING` state, I would think it would already have been indexed, in which case the indexer would be uninterested in it...is that right?  (But, I can see a case where there would be a conflict between indexing and deletion; there is also a conflict between publishing and deletion, but I think you indicated that that is addressed in a different way...I've lost my grip on this stuff over the holiday break.  🦃 )

Reply:
> > The `dataset.state` is a _tracking_ mechanism, not a _control_ or synchronization mechanism.
> 
> Understood. But, at least in the immediate term, I think we're going to lean on it as though it were a control mechanism. That is, while I'm pretty sure that there is nothing that we can do to prevent two simultaneous requests, one a "delete" and one a "publish", from proceeding if the dataset is already in `INDEXED` state (unless the Server is single threaded...which, perhaps it is...),

We run multiple `gunicorn` instances (5?), so we can handle simultaneous requests.

I've experimented with transaction control, but the SQLAlchemy semantics seem to be a bit non-obvious, especially when dealing with both PostgreSQL and sqlite3. (E.g., the default session setting is `autobegin=True`, which has SQLAlchemy automatically creating sessions, so that an explicit `session.begin()` causes problems, while nesting a transaction doesn't seem to work consistently across both DB engines. I've previously considered disabling `autobegin`, but I haven't been up to figuring out how that will affect everything we're already doing; and I certainly don't want to get into that here.)

I'd really much prefer to just go with the basic mitigation I've already coded here, and avoid boiling more of the lake than necessary. Yeah, better state management will become important... but that was never the point of this PR, which was just trying to improve the behavior when you tried to delete or publish before indexing.

> if one arrives slightly before the other, so that the second finds the dataset in `DELETING` or `PUBLISHING` state, then we can report the conflict and avoid the collision problem.
> 
> > If I add `States.PUBLISHING`, there's nothing to keep the indexer from pulling a dataset in the "publishing" state
> 
> In order for a dataset to be in the `PUBLISHING` state, I would think it would already have been indexed, in which case the indexer would be uninterested in it...is that right?

Currently, but only because it requires an index map via the new constructor flags; but see [PBENCH-982](https://issues.redhat.com/browse/PBENCH-982). There's really no reason we shouldn't allow publish earlier, or even on `PUT` (e.g., with `?access=public`); it's just that right now the indexer assumes the access is "private" whereas it'd need to be driven by the dataset column.